### PR TITLE
Add support for 'maxTotalRows' to incremental publishing kafka indexing task and appenderator based realtime task

### LINKED ISSUE #5898: Implement Kafka "maxTotalRows" parameter
Kafka indexing has a tunable "maxRowsPerSegment" to allow it to publish data early if a single segment has too much data. This is good, but it doesn't help in a scenario where there is a very large amount of data spread over a large number of segments.

This can happen if you are doing a backfill through Kafka: imagine having segmentGranularity DAY, and a Kafka topic that you are pre-populating with a year's worth of historical data. Perhaps you've set "maxRowsPerSegment" to 5 million, and each day has 4 million rows. By the time you've read a month's worth of data, you've got 120 million rows tied up in the kafka task, and maxRowsPerSegment hasn't been triggered yet. Query speeds slow to a crawl, and it's questionable if publishing and handoff will ever succeed.

This could be fixed by introducing a "maxTotalRows" parameter that applies across _all_ segments, similar to the one that the index task uses: http://druid.io/docs/latest/ingestion/tasks.html.

### PR DESCRIPTION
**Pull request by clintropolis**
Resolves #5898 by adding `getMaxTotalRows` and `getMaxRowsPerSegment` to `AppenderatorConfig`, extending the model used by `IndexTask` to  `IncrementalPublishingKafkaIndexTaskRunner` and `AppenderatorDriverRealtimeTask`. 

Additionally, tweaks `maxRowsPerSegment` behavior of kafka indexing to match the appenderator based realtime indexing change in #6125.

### CODE REVIEW

File: extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/IncrementalPublishingKafkaIndexTaskRunner.java [hunk 1]
```diff
@@ -429,9 +429,6 @@ public void run()
           // if stop is requested or task's end offset is set by call to setEndOffsets method with finish set to true
           if (stopRequested.get() || sequences.get(sequences.size() - 1).isCheckpointed()) {
```
[the remark this answers is not in the document]
Comment (clintropolis) [reply extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/IncrementalPublishingKafkaIndexTaskRunner.java]: I discussed with @jihoonson because I noticed that previously it was setting the state to publishing and then going through a bunch of logic not related to publishing if stop wasn't user requested, and this is what was intended to happen here. As I understand, currently it will harmlessly but needlessly fall through the logic below before breaking out and publishing.

File: server/src/main/java/io/druid/segment/realtime/appenderator/AppenderatorConfig.java [hunk 2]
```diff
@@ -30,12 +30,48 @@
...
+    return Integer.MAX_VALUE;
+  }
+
+  /**
+   * Maximum number of rows across all segments before pushing to deep storage
+   *
+   * @return
+   */
+  @Nullable
+  default Long getMaxTotalRows()
```
[the remark this answers is not in the document]
Comment (clintropolis) [reply server/src/main/java/io/druid/segment/realtime/appenderator/AppenderatorConfig.java]: This should never be used afaik, everything that uses this value gets it from json, this default is so I didn't have to add it to `RealtimeTuningConfig` which implements `AppenderatorConfig`.

File: server/src/main/java/io/druid/segment/realtime/appenderator/AppenderatorImpl.java [hunk 3]
```diff
@@ -611,6 +603,7 @@ public Object doCall() throws IOException
       }
       theSinks.put(identifier, sink);
       sink.finishWriting();
+      totalRows.addAndGet(-sink.getNumRows());
```
[the remark this answers is not in the document]
Comment (clintropolis) [reply server/src/main/java/io/druid/segment/realtime/appenderator/AppenderatorImpl.java]: Hmm, where you are suggesting? Those values are decremented at persist time, this one is happening at publish time.

File: server/src/main/java/io/druid/segment/realtime/appenderator/AppenderatorImpl.java [hunk 4]
```diff
@@ -1118,16 +1094,18 @@ public String apply(SegmentIdentifier input)
       final boolean removeOnDiskData
   )
   {
+    if (sink.isWritable()) {
```
[the remark this answers is not in the document]
Comment (clintropolis) [reply server/src/main/java/io/druid/segment/realtime/appenderator/AppenderatorImpl.java]: For one, it seemed to be more legit to check that to match the `we only count active sinks` comment which implies writable to me. Additionally, since there are a couple of paths to decrementing the `totalRows` counter, I wanted to make sure I never double decremented it (it's other decrement is also tied to finishWriting).

File: server/src/main/java/io/druid/segment/realtime/appenderator/AppenderatorImpl.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (gianm) [reply server/src/main/java/io/druid/segment/realtime/appenderator/AppenderatorImpl.java]: Oh, yeah, you're right. Let me re-read this with non-dumb eyes.

File: server/src/main/java/io/druid/segment/realtime/appenderator/AppenderatorConfig.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (gianm) [reply server/src/main/java/io/druid/segment/realtime/appenderator/AppenderatorConfig.java]: If it's never going to be used, how about throwing UnsupportedOperationException and marking it non-nullable?

[the remark this answers is not in the document]
Comment (clintropolis) [reply server/src/main/java/io/druid/segment/realtime/appenderator/AppenderatorConfig.java]: It's nullable to be consistent with [how `IndexTask` was using it](https://github.com/apache/incubator-druid/blob/master/indexing-service/src/main/java/io/druid/indexing/common/task/IndexTask.java#L1106). I don't have strong opinions about `null` vs `Long.MAX_VALUE`, I'll rework where this is used to get rid of nullable.

File: server/src/main/java/io/druid/segment/realtime/appenderator/AppenderatorImpl.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (gianm) [reply server/src/main/java/io/druid/segment/realtime/appenderator/AppenderatorImpl.java]: Hmm, to prevent races (abandonSegment _could_ be called in a separate thread) how about having `sink.finishWriting()` return true or false, corresponding to its writability state before the method was called? Then, only decrement the counters if it returns true.

File: server/src/main/java/io/druid/segment/realtime/appenderator/AppenderatorImpl.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (gianm) [reply server/src/main/java/io/druid/segment/realtime/appenderator/AppenderatorImpl.java]: Ok, I re-read it, less dumbly this time, and it looks good to me. But consider the finishWriting-returning-boolean thing.

File: server/src/main/java/io/druid/segment/realtime/appenderator/AppenderatorConfig.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (clintropolis) [reply server/src/main/java/io/druid/segment/realtime/appenderator/AppenderatorConfig.java]: Oh, I didn't read the comment on that, it looks like `IndexTask` needs this to be nullable, will `throw UnsupportedOperationException` in default method at least.

File: server/src/main/java/io/druid/segment/realtime/appenderator/AppenderatorImpl.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (clintropolis) [reply server/src/main/java/io/druid/segment/realtime/appenderator/AppenderatorImpl.java]: :+1:

**Comment by jihoonson** [comment]
I'm reviewing this PR.

File: extensions-core/kafka-indexing-service/src/test/java/io/druid/indexing/kafka/KafkaIndexTaskTest.java [hunk 5]
```diff
@@ -281,6 +282,7 @@ public KafkaIndexTaskTest(boolean isIncrementalHandoffSupported)
     );
   }
 
+
```
Comment (jihoonson) [inline extensions-core/kafka-indexing-service/src/test/java/io/druid/indexing/kafka/KafkaIndexTaskTest.java]: Please remove this.

File: extensions-core/kafka-indexing-service/src/test/java/io/druid/indexing/kafka/KafkaIndexTaskTest.java [hunk 6]
```diff
@@ -476,7 +478,7 @@ public void testIncrementalHandOff() throws Exception
     }
     final String baseSequenceName = "sequence0";
     // as soon as any segment has more than one record, incremental publishing should happen
-    maxRowsPerSegment = 1;
+    maxRowsPerSegment = 2;
```
Comment (jihoonson) [inline extensions-core/kafka-indexing-service/src/test/java/io/druid/indexing/kafka/KafkaIndexTaskTest.java]: @clintropolis would you tell me why this change is needed?

File: extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaTuningConfig.java [hunk 7]
```diff
@@ -40,6 +40,7 @@
   private final int maxRowsInMemory;
   private final long maxBytesInMemory;
   private final int maxRowsPerSegment;
+  private final Long maxTotalRows;
```
Comment (jihoonson) [inline extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaTuningConfig.java]: Please add `@Nullable`.

File: extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaTuningConfig.java [hunk 8]
```diff
@@ -153,12 +157,21 @@ public long getMaxBytesInMemory()
...
   @JsonProperty
   public int getMaxRowsPerSegment()
   {
     return maxRowsPerSegment;
   }
 
+
+  @JsonProperty
+  @Override
+  public Long getMaxTotalRows()
```
Comment (jihoonson) [inline extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaTuningConfig.java]: `@Nullable`.

File: indexing-service/src/main/java/io/druid/indexing/common/index/RealtimeAppenderatorTuningConfig.java [hunk 9]
```diff
@@ -53,8 +53,9 @@ private static File createNewBasePersistDirectory()
   }
 
   private final int maxRowsInMemory;
-  private final int maxRowsPerSegment;
   private final long maxBytesInMemory;
+  private final int maxRowsPerSegment;
+  private final Long maxTotalRows;
```
Comment (jihoonson) [inline indexing-service/src/main/java/io/druid/indexing/common/index/RealtimeAppenderatorTuningConfig.java]: `@Nullable`

File: indexing-service/src/main/java/io/druid/indexing/common/index/RealtimeAppenderatorTuningConfig.java [hunk 10]
```diff
@@ -138,12 +145,20 @@ public long getMaxBytesInMemory()
...
+  @Override
   @JsonProperty
   public int getMaxRowsPerSegment()
   {
     return maxRowsPerSegment;
   }
 
+  @Override
+  @JsonProperty
+  public Long getMaxTotalRows()
```
Comment (jihoonson) [inline indexing-service/src/main/java/io/druid/indexing/common/index/RealtimeAppenderatorTuningConfig.java]: `@Nullable`

File: indexing-service/src/main/java/io/druid/indexing/common/task/IndexTask.java [hunk 11]
```diff
@@ -1485,6 +1485,7 @@ public long getMaxBytesInMemory()
     }
 
     @JsonProperty
+    @Override
```
Comment (jihoonson) [inline indexing-service/src/main/java/io/druid/indexing/common/task/IndexTask.java]: Please add `@Nullable` here too.

File: server/src/main/java/io/druid/segment/realtime/appenderator/AppenderatorImpl.java [hunk 12]
```diff
@@ -1119,15 +1096,16 @@ public String apply(SegmentIdentifier input)
   )
   {
     // Ensure no future writes will be made to this sink.
-    sink.finishWriting();
+    if (sink.finishWriting()) {
+      // Decrement this sink's rows from rowsCurrentlyInMemory (we only count active sinks).
+      rowsCurrentlyInMemory.addAndGet(-sink.getNumRowsInMemory());
```
Comment (jihoonson) [inline server/src/main/java/io/druid/segment/realtime/appenderator/AppenderatorImpl.java]: Hmm, is this valid? It looks that this would not be called if [this](https://github.com/apache/incubator-druid/pull/6129/files#diff-940e5274e2cfb880c70e6249ea864f63R605) is executed first.

File: extensions-core/kafka-indexing-service/src/test/java/io/druid/indexing/kafka/KafkaIndexTaskTest.java (hunk 6 above)
Comment (clintropolis) [reply extensions-core/kafka-indexing-service/src/test/java/io/druid/indexing/kafka/KafkaIndexTaskTest.java]: This PR also matches the behavior of #6125 which changes a '>' to a '>=' and whose logic was moved [here in this PR](https://github.com/apache/incubator-druid/pull/6129/files/923537d4090a049f267f96266ec706b5a7cfdf19#diff-30359ba3286f9021542c0e2ef7738572R105). So to not modify the test a ton i upped the count (since with new behavior the test scenario would be pushing every row).

I updated the main description of the PR to reflect this, thanks for the reminder, I forgot :+1:

File: indexing-service/src/main/java/io/druid/indexing/common/task/IndexTask.java (hunk 11 above)
Comment (clintropolis) [reply indexing-service/src/main/java/io/druid/indexing/common/task/IndexTask.java]: Done (but line changed is below this comment so github isn't squashing this one in the UI)

**Comment by clintropolis** [comment]
@jihoonson do you have any additional comments? I believe I've addressed all existing review.
